driver-core/src/main/com/mongodb/connection/ByteBufPayloadBsonDocument.java

```diff
@@ -0,0 +1,134 @@
+import java.io.Serializable;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.List;
+
+class ByteBufPayloadBsonDocument extends AbstractByteBufBsonDocument implements Cloneable, Serializable {
+    private static final long serialVersionUID = 3L;
+    private static final Charset UTF8_CHARSET = Charset.forName("UTF-8");
+
+    private final ByteBufBsonDocument commandDocument;
+    private final String payloadName;
+    private final List<ByteBufBsonDocument> payload;
+    private BsonDocument unwrapped;
+
+
+    static ByteBufPayloadBsonDocument create(final ByteBufferBsonOutput bsonOutput, final int startPosition) {
+        List<ByteBuf> duplicateByteBuffers = bsonOutput.getByteBuffers();
+        CompositeByteBuf outputByteBuf = new CompositeByteBuf(duplicateByteBuffers);
+        outputByteBuf.position(startPosition);
+
+        int curDocumentStartPosition = startPosition;
+        int documentSizeInBytes = outputByteBuf.getInt();
+        ByteBuf slice = outputByteBuf.duplicate();
+        slice.position(curDocumentStartPosition);
+        slice.limit(curDocumentStartPosition + documentSizeInBytes);
+        ByteBufBsonDocument commandDocument = new ByteBufBsonDocument(slice);
+
+        curDocumentStartPosition += documentSizeInBytes;
+
+        slice = outputByteBuf.duplicate();
+        slice.position(curDocumentStartPosition);
+        if (slice.get() != 1) {
+            throw new IllegalArgumentException("Invalid payload type");
+        }
+        int sliceSize = slice.remaining();
+        int payloadSize = slice.getInt();
+        if (sliceSize != payloadSize) {
+            throw new IllegalArgumentException("Invalid payload size");
+        }
+
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        byte nextByte = slice.get();
+        while (nextByte != 0) {
+            outputStream.write(nextByte);
+            nextByte = slice.get();
+        }
+        String payloadName = new String(outputStream.toByteArray(), UTF8_CHARSET);
+
+        List<ByteBufBsonDocument> payload = new ArrayList<ByteBufBsonDocument>();
+        curDocumentStartPosition = slice.position();
+        while (outputByteBuf.hasRemaining()) {
+            documentSizeInBytes = slice.getInt();
+            slice.position(curDocumentStartPosition);
+            slice.limit(curDocumentStartPosition + documentSizeInBytes);
+            ByteBufBsonDocument doc = new ByteBufBsonDocument(slice);
+            payload.add(doc);
+            curDocumentStartPosition += documentSizeInBytes;
+            outputByteBuf.position(curDocumentStartPosition);
+        }
+
+        for (ByteBuf byteBuffer : duplicateByteBuffers) {
+            byteBuffer.release();
+        }
+        return new ByteBufPayloadBsonDocument(commandDocument, payloadName, payload);
+    }
+
+    <T> T findInDocument(final Finder<T> finder) {
+        BsonDocumentReader bsonReader = new BsonDocumentReader(toBsonDocument());
```

Comment:
This is suspect.  Above we go through a lot of effort not to hydrate any documents unnecessarily, but then whenever anyone needs to look inside for anything, we hydrate the whole thing.  And this will happen every time due to the need to look for the "$query" special key in InternalStreamConnection.CommandEventSender#sendStartedEvent.

Could we optimize this by first looking in the main command document, and then looking in the payload?

Reply:
Added JAVA-2636 to track that improvement.